compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertToDma.cpp

```diff
@@ -146,129 +154,400 @@ LogicalResult unPackDmaInputs(IREE::LinalgExt::UnPackOp unPackOp,
+  builder.setInsertionPoint(subviewOp);
+  offsets = getIndexOpFoldResultSum(builder, subviewOp.getLoc(), offsets,
+                                    subviewOp.getMixedOffsets());
+
+  sizes = subviewOp.getMixedSizes();
+  if (llvm::any_of(sizes, [](OpFoldResult size) {
+        return !getConstantIntValue(size).has_value();
+      })) {
+    return subviewOp->emitOpError(
+        "has dynamic shape that is not supported by the target dma op.");
+  }
+
+  assert(offsets.size() == sizes.size() && sizes.size() == strides.size() &&
+         "mismatch in the number of offsets, sizes and strides");
+
+  // Handle the case where some dimensions are dropped in the subview:
+  llvm::SmallBitVector droppedDims = subviewOp.getDroppedDims();
+  uint64_t insertionIndex{0};
+  for (uint64_t extractionIndex = 0; extractionIndex < offsets.size();
+       ++extractionIndex) {
+    if (!droppedDims[extractionIndex]) {
+      offsets[insertionIndex] = offsets[extractionIndex];
+      sizes[insertionIndex] = sizes[extractionIndex];
+      strides[insertionIndex] = strides[extractionIndex];
+      insertionIndex++;
+    }
+  }
+  offsets.resize(insertionIndex);
+  sizes.resize(insertionIndex);
+  strides.resize(insertionIndex);
+  return success();
+}
+
+/// Provide the offsets, sizes and strides of the inputs to `operandOp`.
+/// This function updates `operandOp`, setting it to the allocation operation
+/// that it originates from.
+LogicalResult setDmaInputs(Operation *&operandOp,
+                           SmallVector<OpFoldResult> &offsets,
+                           SmallVector<OpFoldResult> &sizes,
+                           SmallVector<OpFoldResult> &strides) {
+  assert(offsets.empty() && sizes.empty() && strides.empty() &&
+         "offsets, sizes, and strides must be empty");
+
+  if (!operandOp) assert(false && "operandOp must be non-null");
+
+  // Get the sequence of memref operations going from an allocation to
+  // `operandOp`
+  SmallVector<Operation *> chain;
+
+  Operation *currentOp = operandOp;
+  while (currentOp) {
+    chain.push_back(currentOp);
+    if (isAllocation(currentOp)) {
+      currentOp = {};
+    } else if (auto memref = dyn_cast<memref::SubViewOp>(currentOp)) {
+      currentOp = memref.getSource().getDefiningOp();
+    } else if (auto expand = dyn_cast<memref::ExpandShapeOp>(currentOp)) {
+      currentOp = expand.getSrc().getDefiningOp();
+    } else if (auto collapse = dyn_cast<memref::CollapseShapeOp>(currentOp)) {
+      currentOp = collapse.getSrc().getDefiningOp();
+    } else {
+      return currentOp->emitOpError("is currently not supported");
+    }
+  }
+
+  operandOp = chain.back();
+  if (!isAllocation(operandOp)) {
+    return operandOp->emitOpError(
+        "is not a memref.alloc or hal.interface.binding.subspan operation.");
+  }
+
+  // Starting from the allocation, update the offsets, sizes, and strides.
+  for (auto iter = chain.rbegin(); iter != chain.rend(); ++iter) {
```

Comment:
I'm okay with your method first going up the chain to find the definingOp and then going down the chain to update the address. But I think a recursion could make the codes cleaner.